src/ir/table-utils.h

```diff
@@ -67,10 +67,23 @@ inline Table::Segment& ensureTableWithOneSegment(Table& table, Module& wasm) {
 // Appends a name to the table. This assumes the table has 0 or 1 segments,
 // as with 2 or more it's ambiguous what we should do (use a hole in the middle
 // or not).
+// This works on code from wasm-ld, but on arbitrary code it may not be valid
+// in the presence of a dynamic linking section. Specifically, we assume the
+// module has a single table segment, and that the dylink section indicates
+// we can validly append to that segment, see the check below.
 inline Index append(Table& table, Name name, Module& wasm) {
   auto& segment = ensureTableWithOneSegment(table, wasm);
-  table.segments[0];
   auto tableIndex = segment.data.size();
+  if (wasm.dylinkSection) {
+    if (segment.data.size() != wasm.dylinkSection->tableSize) {
+      Fatal() << "Appending to the table in a module with a dylink section "
+                 "that has tableSize which indicates it wants to reserve more "
+                 "table space than the actual table elements in the module. "
+                 "We don't know how to correctly update the dylink section in "
+                 "that case.";
+    }
```

Comment:
What do you think about about making this assertion when validating / writing the the module ?

Reply:
I don't think that would be ok for us to do. It would mean we can't roundtrip valid dynamic libraries (valid according to wasm and the dynamic linking conventions).

Reply:
Thats what I mean, isn't it the case that there might be binaryen passed that currently break such validation.   E.g. anyone that adds or removes a table entry but forgets to update the dylink sections has broken validation no?  Don't we want to warn when this happens?

Reply:
I see what you mean about verifying we don't have bad passes, but it seems like we'd need to save some state on load (initial state of the table, initial state of the dylink section) and compare those when writing. That seems potentially useful but I'm not sure it's worth adding global state for it.

And specifically, we don't have any passes that do anything like that AFAIK. If we did they should be using the central `TableUtils` code where we have a check now.